### PR TITLE
updated shutdownDrainTarget in BatchExportProcessor to be thread-safe

### PR DESCRIPTION
**Pull request by Yun-Ting**
Related to https://github.com/open-telemetry/opentelemetry-dotnet/issues/4226

Design discussion issue #

## Changes

Updated `shutdownDrainTarget` to use `Volatile.Read()` and `Volatile.Write()` to make sure the operations are atomic in 32-bit systems. Theoretically, this won't be a problem, because https://github.com/open-telemetry/opentelemetry-dotnet/blob/9395b9e944cdae19a9823c0bc3dacb257c9c4863/src/OpenTelemetry/BaseProcessor.cs#L112 guarantees that only one thread will enter `.OnShutdown` at a time, add Volatile though to ensure thread safety if this dependency is broken.

Note: In x64 system,  `Volatile.Read()` and `Volatile.Write()` will be optimized away. https://github.com/dotnet/runtime/blob/e247ef3c03f5dae7ed634d83ec53e1386e5fe57d/src/libraries/System.Private.CoreLib/src/System/Threading/Volatile.cs#L87-L107

x86 assembly before adding Volatile:
image
[SharpLab](https://sharplab.io/#v2:EYLgdgpgLgZgHgGiiAhgZwLYB8ACAmARgFgAoHABgAIcCA6AFQAsAnCFAEwEswBzAblKkcAZmp5KAYU7MAxgFcANimYAhOTBgRmpAN6lKBgA7NOANxRQIlBQHtelRm3YCSBysbMWrt+1BScFFzdKfQMRazseSgBBdnYIdgkbOTAoSgBeAD5KADUbJSgAiFoAJScAClYYSihGTjRaRw4ASj5g0IMO6lEfKLKMG1MEpJS0rNz8iyLSiqqauoa/ANauruDw7jSR1LW3PVdgw4MeaF2j/aPLt3NmGv8FAGUwFEM0RhsxiYLpso5KiGqtXqtCWChWByuRxwAHZKOVNs1ynlvgpir92P9AQtGk5mpQALR3AJPF5vD7gyFuAC+Z0oVPaJBpJCEonwlBUFhkjAAonBDDZmFAAArMGwyCBoNACygXS4ecyWSisDh2BQAT0k0nkSlU6k0txkWsUyjUGi0QTlJgV3kilDecig7BsAHcwAARZj+MD0ZQnT69WgAWRQcByKAUcggFs6EMO4Q5UC5vP5gpFYolUuY5WatNllMNsmNurNt3SlEgzs1hZ1pv12ejhyZDKu4WANnylAA8mAHowHU7XfDUjVOBgIMkoIGAgp6hAZHZ2Ggc7HgnnIUCGvbHS73Z7uD7mH6MvNgQXtSa9VpaLF4olkqkG1daetYVBmJHH9TViR5V5qAAWShkwFSxmDTGRs10LpnTqVE4TfSNl1XWlOGqcoN1oM8i1rK9+kGYZ7zSTIywwrcB13L0Dz9JDLjXFtoU/Awmy/RkGSpIA===)

x86 assembly after adding Volatile
image
[SharpLab](https://sharplab.io/#v2:EYLgdgpgLgZgHgGiiAhgZwLYB8ACAmARgFgAoHABgAIcCA6AFQAsAnCFAEwEswBzAblKkcAZmp5KAYU7MAxgFcANimYAhOTBgRmpAN6lKBgA7NOANxRQIlBQHtelRm3YCSBysbMWrt+1BScFFzdKfQMRazseSgBBdnYIdgkbOTAoSgBeAD5KADUbJSgAiFoAJScAClYYSihGTjRaRw4ASj5g0IMO6lEfKLKMG1MEpJS0rNz8iyLSiqqauoa/ANauruDw7jSR1LW3PVdgw4MeaF2j/aPLt3NmGv8FAGUwFEM0RhsxiYLpso5KiGqtXqtCWChWByuRxwAHZKOVNs1ynlvgpir92P9AQtGk5mpQALR3AJPF5vD7gyFuAC+Z0oVPaJBpJCEonwlBUFhkjAAonBDDZmFAAArMGwyCBoNACygXS4ecyWSisDh2BQAT0k0nkSlU6k0txkWsUyjUGi0QTlJgV3kilDecig7BsAHcwAARZj+MD0ZQnT69WgAWRQcByKAUcggFs6EMO4Q5UC5vP5gpFYolUuY5WatNllMNsmNurNt3SlEgzs1hZ1pv12ejhyZDKu4WANnylAA8mAHowHU7XfDUjVOBgIMkoIGAgp6hAZHZ2Ggc7HgnnIcipqjaAB1EyWTHzYH2x0u92e7g+5h+hCHhoF7UmvVaWixeKJZKpCmUkIry4wmrMJGDbBE2bikPKXjUAALJQyYCpYzBpjI2a6F0zp1KicJQIBEDLqutKcNU5RAneRo1k+zAzAMQzvqMlCZGWG6FFu6IHiRtDHgOZ5epefrNHhlxri20LAdSXRMvSYGMkAA=)

Volatile will be optimized away for x64
image
[SharpLab](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAmAjAWAChMAGAAk1wDoAVACwCcBTRAEwEsA7AcwG4iimAMwVsZAMLsGAYwCuAG0QMAQrIBmapgyIBvImQMAHBuwBuiOEzLyA9jzJ0WrfoQNljZi1dv24idvIubmT6BsLWdtxkAIKsrEys4jaynHBkALwAfGQAajaKcAFMVABKTgAUzGpkcHTsyFSObACUvMGhBh0UIj5RZag2pglJKWlZufkWRaUVVTV1DX4BrV1dweFcaSOpa256rsGHBtxMcLtH+0dXbuYMNf7yAMqciIbIdDZjEwXTZWyVTGqtXqVCW8hWB2uR0wAHYyOVNs1ynkfvJin9WACgQtGk5mmQALT3ALPV7vT4QqFuAC+5zI1PahFphEEIhwZGUFmkdAAomBDDYGHAAAoMGzSJjIZCCsiXK4ecyWMjMNh2eQATwkUjkihU6k0d2k2oUSlUGi0QXlJkV3kiZHesjgrBsAHdOAARBj+Tg0JQnL69KgAWUQYByiHksiYls6kMO4U5cG5fIFQtF4sl0oY5WadLlVKNMhNevNd3SZE4TBdWqLurNBpzMcOzMZ13CwBs+TIAHlOI86I7nW6EakauxUExknAgwF5PUmNI7KxkLm48F81CUVM0VQAOomSxY+Ygh1O10er1cX0Mf0wY8NQs6036rRUWLxRLJVKUqkhNdXWEagYKMm2CFs3CIBUvAoKAyBTQVLAYdNpBzXQuhdOo0XhOBgKYVd1zpdhqnKYEH2NOsXwYGYBiGT9RjITJyy3QodwxI9SKoU8hwvb1r39Zp8KuDc2xhUCaS6ZkGQgpkgA===)

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Review by CodeBlanch — APPROVED** [review]
LGTM

File: src/OpenTelemetry/BatchExportProcessor.cs [hunk 1]
```diff
@@ -85,6 +85,8 @@ public abstract class BatchExportProcessor<T> : BaseExportProcessor<T>
         /// </summary>
         internal long DroppedCount => this.droppedCount;
 
+        internal long ShutdownDrainTarget => Volatile.Read(ref this.shutdownDrainTarget);
```
Comment (utpilla) [inline src/OpenTelemetry/BatchExportProcessor.cs:88]: This property is not required.

Comment (Yun-Ting) [reply src/OpenTelemetry/BatchExportProcessor.cs]: removed

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry/BatchExportProcessor.cs [hunk 2]
```diff
@@ -198,7 +198,7 @@ protected override bool OnForceFlush(int timeoutMilliseconds)
                     return true;
                 }
 
-                if (this.shutdownDrainTarget != long.MaxValue)
+                if (Volatile.Read(ref this.shutdownDrainTarget) != long.MaxValue)
```
Comment (ThomsonTan) [inline src/OpenTelemetry/BatchExportProcessor.cs:201]: Wondering why do we use `Volatile.Read`/`Volatile.Write` to get atomic semantic. As per the doc below, both functions are for consumed by compiler to decide inserting memory barrier or not, and the `InterLocked` instruction could have the side-effect of inserting barrier. I also looked at the implementation in the runtime shared here, which says 32-bit ordinary read is not atomic so lock instruction required. Not sure if it implies that ordinary read of 32-bit is atomic on x64, or even on ARM64, the behavior may be very different and could still required explicit lock.

https://learn.microsoft.com/en-us/dotnet/api/system.threading.volatile.read?view=net-8.0

**Pull request merged**